### PR TITLE
DPA-7217 Parameter validation HOFLER

### PR DESCRIPTION
**Pull request by BukinStanislav**
https://jira-xa.x-tensive.com/browse/DPA-7217

### CODE REVIEW

File: Parameters/HOFLER/Creation/HOFLER_PARAMETER_CREATION_TRIGGER.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+        private void Watcher_Created(object sender, FileSystemEventArgs e)
+        {
+            OnSignal(e.FullPath);
+        }
+
+        public override Task StopAsync()
+        {
+            watcher.StopWatch();
+            return Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/Creation/HOFLER_PARAMETER_CREATION_TRIGGER.cs]: ! done

File: Parameters/HOFLER/Validation/HOFLER_PARAMETER_VALIDATION_TRIGGER.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+        private void Watcher_Created(object sender, FileSystemEventArgs e)
+        {
+            OnSignal(e.FullPath);
+        }
+
+        public override Task StopAsync()
+        {
+            watcher.StopWatch();
+            return Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/Validation/HOFLER_PARAMETER_VALIDATION_TRIGGER.cs]: done

File: Parameters/HOFLER/Common/HOFLER_PARAMETER_CONFIG.cs [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+                .ToArray();
+
+            validationModel.ControlProgramName = controlProgramName;
+
+            var controlsInfo = jsonModel.Dialogs
+                .Where(dialog => !IgnoredDialogs.Contains(dialog.ID))
+                .SelectMany(dialog => dialog.Controls
+                    .Where(control => !IgnoredControls.Contains(control.ID))
+                    .Select(control => new { Control = control, DialogName = dialog.Name }));
+            var numberItem = "No";
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/Common/HOFLER_PARAMETER_CONFIG.cs]: done

File: Parameters/HOFLER/Common/HOFLER_PARAMETER_CONFIG.cs [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+
+            validationModel.WorkCenterName = workCenter.Name;
+            validationModel.Channel = channel;
+
+            var parameterSets = Query.All<VersionOfControlProgram>()
+                .Where(version => version.VersionStatus == Status.Active && version.ControlProgram.Name == controlProgramName)
+                .SelectMany(version => version.LinkedParameterSets.Select(parameterSetLink => parameterSetLink.ParametersSet)
+                    .Union(version.LinkedSubprograms.SelectMany(subprogramLink => subprogramLink.ParametersSetLinks
+                        .Select(parameterSetLink => parameterSetLink.ParametersSet))))
+                .ToArray();
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/Common/HOFLER_PARAMETER_CONFIG.cs]: refactored and split into several methods

File: Parameters/HOFLER/Validation/HOFLER_PARAMETER_VALIDATION_HANDLER.cs [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+        }
+
+        private void WriteFile(string directory, string fileName, string content)
+        {
+            if (!Directory.Exists(directory)) {
+                Directory.CreateDirectory(directory);
+            }
+
+            var targetFilePath = Path.Combine(directory, fileName);
+            System.IO.File.WriteAllText(targetFilePath, content);
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/Validation/HOFLER_PARAMETER_VALIDATION_HANDLER.cs]: done

File: Parameters/HOFLER/StockData/HOFLER_STOCKDATA_TRIGGER.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+        private void Watcher_Created(object sender, FileSystemEventArgs e)
+        {
+            OnSignal(e.FullPath);
+        }
+
+        public override Task StopAsync()
+        {
+            watcher.StopWatch();
+            return Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/StockData/HOFLER_STOCKDATA_TRIGGER.cs]: done

File: Parameters/HOFLER/PDCLog/HOFLER_PDCLOG_HANDLER.cs [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+        private static bool Contains(string line, string value)
+        {
+            return line.IndexOf(value, StringComparison.Ordinal) > -1;
+        }
+
+        public PDCLogState GetState(string line)
+        {
+            state = rules[state](line);
+            return state;
+        }
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/PDCLog/HOFLER_PDCLOG_HANDLER.cs]: refactored state machine

File: Parameters/HOFLER/Archiving/HOFLER_ARCHIVING_TRIGGER.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    public class HOFLER_ARCHIVING_TRIGGER : Signals2TriggerBase
+    {
+        private TimeTrigger timeTrigger = null;
+
+        public override Task StartAsync()
+        {
+            timeTrigger?.Dispose();
+
+            var startTime = HOFLER_PARAMETER_CONFIG.ARCHIVING.START_TIME;
+            var now = DateTimeOffset.Now;
```
[the remark this answers is not in the document]
Comment (BukinStanislav) [reply Parameters/HOFLER/Archiving/HOFLER_ARCHIVING_TRIGGER.cs]: done

**Comment by BukinStanislav** [comment]
@VasilyUstugov do not forget to track working time

**Review by VasilyUstugov — APPROVED** [review]

**Pull request merged**